File: Frontend/lib/models/message.dart
class Message {
  final int id;
  final String content;
  final String messageType;
  final String? sentAt;
  final int senderId;
  final String senderName;
  final int? receiverId;
  final int? projectId;

  Message({
    required this.id,
    required this.content,
    required this.messageType,
    this.sentAt,
    required this.senderId,
    required this.senderName,
    this.receiverId,
    this.projectId,
  });

  factory Message.fromJson(Map<String, dynamic> json) => Message(
        id: json['id'],
        content: json['content'],
        messageType: json['messageType'] ?? 'DIRECT',
        sentAt: json['sentAt'],
        senderId: json['senderId'],
        senderName: json['senderName'] ?? '',
        receiverId: json['receiverId'],
        projectId: json['projectId'],
      );

  Map<String, dynamic> toJson() => {
        'content': content,
        'messageType': messageType,
        if (receiverId != null) 'receiverId': receiverId,
        if (projectId != null) 'projectId': projectId,
      };
}

File: Backend/src/main/java/com/hsd/entities/Message.java
package com.hsd.entities;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "message", schema = "hsd")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Message {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "content", columnDefinition = "TEXT", nullable = false)
    private String content;

    @Enumerated(EnumType.STRING)
    @Column(name = "message_type", nullable = false)
    private MessageType messageType;

    @Column(name = "sent_at")
    private LocalDateTime sentAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sender_id", nullable = false)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private User sender;

    /** Nullable: only set when messageType = DIRECT */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "receiver_id")
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private User receiver;

    /** Nullable: only set when messageType = PROJECT */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id")
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private Project project;

    @PrePersist
    protected void onCreate() {
        sentAt = LocalDateTime.now();
    }

    public enum MessageType {
        DIRECT, PROJECT
    }
}
